### ISSUE TITLE
UI / Protocol for dynamic audio pipeline tunables at runtime

### ISSUE THREAD

**Issue by allquixotic** [mopidy/mopidy#1581]
I'm the author of [rbpitch](http://launchpad.net/rbpitch), a plugin for Rhythmbox that exposes a GUI for changing the pitch/speed/tempo of music on the fly, using Gstreamer's `gst-plugins-bad` plugin, `pitch` (from [SoundTouch](http://www.surina.net/soundtouch/)). Despite the package it's in, libsoundtouch produces extremely *high-quality* audio while shifting (independently) the pitch, tempo, and speed of the audio.

**My ultimate goal is to enable something like rbpitch for mopidy.** I can already prove the concept is viable, by changing the `output = ` property to include `pitch` element (with one or more of the pitch/tempo/speed properties set to a value other than 1.0) before passing the audio to the sink. It works fine and the audio is transformed as intended.

The problem is that the value for the pitch/tempo/speed has to be hard-coded in the config file. What I'd like to do is expose the properties as tunables that the end-user can interact with (via sliders, like rbpitch). The UI would of course be frontend-specific, but the protocol would have to expose the tunables somehow.

However, in thinking about this, it seems like this could be a more general protocol enhancement to Mopidy, rather than being *specifically* (and only) about pitch/speed/tempo. An equalizer could be added. Maybe volume could be subsumed in a more generic tunables system. Perhaps there could be a runtime toggle for volume normalization by inserting or removing a normalization element.

There are plenty of use cases I can think of where you would want to either (1) add or remove an element to the Gstreamer pipeline at runtime, or (2) tweak properties of an element at runtime. In many cases, you don't even have to pause playback -- for instance, with the pitch element, you can set values of pitch/speed/tempo while it's in the `PLAYING` state, and the audio stream will adjust very elegantly without so much as a hiccup.

I'm not sure if you want to make this something that plugins can do (or whether plugins can *already* do this), or if the core first needs to be refactored to be more flexible and make this possible.

Basically, the high level architecture of a *generic* system of this sort would involve:

 - On the backend: introduce a system for defining dynamic (runtime) tunables, what kind of data is involved (boolean, int, double, etc), the "friendly name" of the tunable as exposed to the user layer, and the gstreamer-specific "guts" that actually performs the modification of the tunable at runtime.

 - On the frontend: query the system for a list of tunables, get metadata about each tunable, and expose them in the UI based on the type of tunable along with the name, min/max values, etc.

I think a generic system like this would be more generally useful for the future, even though in principle I could hack together some kind of pitch-specific signaling that would be pretty brittle and completely specific to the pitch element.

Although I *would* like to see the (optional) functionality of inserting the `pitch` element into the pipeline, go into mopidy core instead of an out-of-tree plugin...

Thoughts on the design/approach for this? Are some of my ideas that involve core protocol changes already in the works / already implemented so that I could start building off of them?

NB: My developer knowledge of gstreamer is fairly extensive (enough to write rbpitch from scratch in Vala), and I'm familiar with the basic syntax of Python, but I'm no expert on Mopidy's internals (yet).

I would appreciate advice as to how I could help contribute towards this effort so that, eventually, mopidy has an "rbpitch equivalent".

This issue is limited in scope to the mopidy backend / core. For e.g. web clients to adapt to whatever changes are made here, we'll need separate issues in the appropriate github repos.

**Comment by adamcik** [comment]
Skimming over quickly the following issues come to mind:

- #1100 which covers letting extensions expose APIs to other extensions.
- #1123 for equalizer support.
- #1118 for dynamic / plugable / switchable outputs

Though it sounds like we could expand this to have plugable audio transforms in addition to ouputs. Then possibly convert the current software volume support to a "transform". And combine with #1100 for configuring things (though that might not be applicable for the volume case).

So in short I think you are heading in the right idea for adding this. I'll try and followup in more detail when I have some more time to spare.
